MTV-2930 | Prevent mixed copy methods in migration plans

amitosw15 opened the pull request:
## Description
Add validation before plan execution to ensure all disks use a consistent copy method (either VDDK or XCOPY/Offload).

## Logic
- **Multi-VM Plans:** If one VM cannot use offload while others can, the plan must be split (all VMs in a plan must share the copy method).
- **Multi-Disk VMs:** If a VM has multiple disks and one does not support offload, the system requires VDDK for all disks attached to that VM.

Resolves: MTV-2930

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1837,3 +1842,83 @@ func (r *Reconciler) IsValidTargetName(targetName string) error {
+					mapping = m
+					break
+				}
+			}
```
NP: please move this into separate function to get the mapping

rgolangh commented on pkg/controller/plan/validation.go:
```diff
@@ -1837,3 +1842,83 @@ func (r *Reconciler) IsValidTargetName(targetName string) error {
+		Status:   True,
+		Reason:   NotSupported,
+		Category: api.CategoryCritical,
+		Message:  "VM has mixed disk migration methods (VDDK and copy offload). All disks in a VM must use the same migration method.",
```
Users likely not understand immediately that the method is driven by the storage map. Lets help them by referring them to that storage map

rgolangh commented on pkg/controller/plan/validation.go:
```diff
@@ -1837,3 +1842,83 @@ func (r *Reconciler) IsValidTargetName(targetName string) error {
+			}
+		}
+
+		if hasVddkDisks && hasOffloadDisks {
```
This early return is a missed opportunity to return an aggregated message with list of affected VMs. Let the users know what they need to address

rgolangh commented:
Any update on this?

amitosw15 commented:
> Any update on this?

Yep, sorry, I dealt with other issues first. I pushed a new version addressing your comments, which was tested with the following cases:


**should pass:**
1.  one pure vddk
2.  many pure vddk
3.  one pure offload
4.  many pure offload

**should fail:**
1.  one mixed vm
2.  mix of pure vddk and pure offload
3.  mix of mixed and pures

mixed == VM with both offload and VDDK disks

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1843,3 +1864,205 @@ func (r *Reconciler) IsValidTargetName(targetName string) error {
+			Reason:   NotSupported,
+			Category: api.CategoryCritical,
+			Message:  message,
+			Items:    items,
```
Items represents the list of VM references, UI takes this list and shows to which of the VMs the condition applies to. Pleasetake a look at other validations.

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1843,3 +1864,205 @@ func (r *Reconciler) IsValidTargetName(targetName string) error {
+				pureVMs = pureOffloadVms
+				message = basicMessage + " The following VMs are planned to use copy offload: " + strings.Join(pureOffloadVms, ", ") +
+					". The following VMs are planned to use mixed usage of VDDK and copy offload: " + strings.Join(mixedVms, ", ") + "."
+			}
```
Lets just keep the message constant, something that MTV does not supporte mixed copy methods. WE don't need to show the whole list in the message. Imagine that it will scale to 1000 VMs.

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1169,6 +1220,21 @@ func (r *Reconciler) validateVM(plan *api.Plan) error {
+		if mixedUsage {
+			vddkAndOffloadMixedUsage.Items = append(vddkAndOffloadMixedUsage.Items, mixedVms...)
+			vddkAndOffloadMixedUsage.Items = append(vddkAndOffloadMixedUsage.Items, pureVddkVms...)
+			vddkAndOffloadMixedUsage.Items = append(vddkAndOffloadMixedUsage.Items, pureOffloadVms...)
```
It's condition showing the mixed VMs we should not show the pure vddk or pure offload

amitosw15 replied:
But if we have a plan with one vm with offload and one vm with vddk we'' have the same error
Any combination of vddk & offload causes this error

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -855,6 +889,23 @@ func (r *Reconciler) validateVM(plan *api.Plan) error {
+					} else if curVMHasOffload {
+						pureOffloadVms = append(pureOffloadVms, ref.String())
+					} else if curVMHasVddk {
+						pureVddkVms = append(pureVddkVms, ref.String())
```
Please add the logic of adding the `vddkAndOffloadMixedUsage.Items` here, just trying to keep it as consistent as possible

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1870,3 +1936,39 @@ func (r *Reconciler) validateConversionTempStorage(plan *api.Plan) error {
+	}
+
+	return diskDetails, hasOffloadDisks, hasVddkDisks, nil
+}
```
We need to check if thje VM has mixed disks. Think this should return just on bool.

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1870,3 +1920,57 @@ func (r *Reconciler) validateConversionTempStorage(plan *api.Plan) error {
+		}
+	}
+	return false
+}
```
No need for this, we already have
```
		plan.IsUsingOffloadPlugin()
```

mnecas commented on pkg/controller/plan/validation.go:
```diff
@@ -1870,3 +1920,57 @@ func (r *Reconciler) validateConversionTempStorage(plan *api.Plan) error {
+		}
+	}
+	return nil
+}
```
Think this can be replaced by 
```
storageMap.FindStorage(datastoreID)
```

amitosw15 commented:
/backport release-2.10
